.ddev/ci/matrix.py

```diff
@@ -0,0 +1,190 @@
+            ['git', *args], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, encoding='utf-8', check=True
+        )
+    except subprocess.CalledProcessError as e:
+        raise OSError(f'{str(e)[:-1]}:\n{e.output}') from None
+
+    return process.stdout
+
+
+def get_changed_files(*, ref: str, local: bool) -> list[str]:
+    changed_files = set()
+
+    # Committed e.g.:
+    # A   relative/path/to/file.added
+    # M   relative/path/to/file.modified
+    for line in git('diff', '--name-status', f'{ref}...').splitlines():
+        if not is_git_warning_line(line):
+            changed_files.add(line.split(maxsplit=1)[1])
+
+    if local:
+        # Tracked
+        for line in git('diff', '--name-only', 'HEAD').splitlines():
+            if not is_git_warning_line(line):
+                changed_files.add(line)
+
+        # Untracked
+        changed_files.update(git('ls-files', '--others', '--exclude-standard').splitlines())
+
+    return sorted(changed_files, key=lambda relative_path: (relative_path, -relative_path.count('/')))
+
+
+def main():
+    root = Path.cwd()
+    os.chdir(root)
+
+    parser = argparse.ArgumentParser(prog='devenv', allow_abbrev=False)
+    parser.add_argument('--ref', default='origin/master')
+    parser.add_argument('-p', '--pretty', action='store_true')
+    parser.add_argument('-v', '--verbose', action='store_true')
+
+    args = parser.parse_args()
+
+    changed_files = get_changed_files(ref=args.ref, local=args.pretty)
+    if args.verbose:
+        print('\n'.join(changed_files), file=sys.stderr)
+
+    changed_directories = {}
+    for changed_file in changed_files:
+        directory_name, _, relative_path = changed_file.partition('/')
+        if relative_path:
+            changed_directories.setdefault(directory_name, []).append(relative_path)
+
+    testable_directories = []
+    for directory_name, files in changed_directories.items():
+        directory = root / directory_name
+        if not ((directory / 'hatch.toml').is_file() and (directory / 'tests').is_dir()):
+            continue
+
+        for relative_path in files:
+            possible_file = directory / relative_path
+            if possible_file.name in NON_TESTABLE_FILES:
+                continue
+            elif any(fnmatch(str(possible_file), str(directory / pattern)) for pattern in TESTABLE_FILE_PATTERNS):
+                testable_directories.append(directory_name)
+                break
+
+    testable_directories.sort(key=lambda d: (DISPLAY_ORDER_OVERRIDE.get(d, len(DISPLAY_ORDER_OVERRIDE)), d))
+
+    overrides = {}
+    if (config_file := root / '.ddev' / 'config.toml').is_file():
+        overrides.update(tomllib.loads(config_file.read_text(encoding='utf-8')).get('overrides', {}))
+
+    display_overrides = overrides.get('display-name', {})
+    ci_overrides = overrides.get('ci', {})
+
+    matrices = []
+    for directory_name in testable_directories:
+        manifest = read_manifest(root, directory_name)
```

Comment:
No this just performs a simple check for documented support of platforms. Occasionally support differs from test environments because locally we may actually have a way to test

Reply:
Gotcha! I also realized that these are 2 different dimensions: hatch mostly talks about host Python/application versions, while manifest talks about host OS.